flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/common/CommonExecSink.java

```diff
@@ -280,26 +297,29 @@ private int deriveSinkParallelism(
      * messages.
      */
     private Transformation<RowData> applyKeyBy(
-            ChangelogMode changelogMode,
+            TableConfig config,
             Transformation<RowData> inputTransform,
             int[] primaryKeys,
             int sinkParallelism,
-            boolean upsertMaterialize) {
-        final int inputParallelism = inputTransform.getParallelism();
-        if ((inputParallelism == sinkParallelism || changelogMode.containsOnly(RowKind.INSERT))
-                && !upsertMaterialize) {
-            return inputTransform;
+            int inputParallelism,
+            boolean inputInsertOnly,
+            boolean needMaterialize) {
+        final ExecutionConfigOptions.SinkKeyedShuffle sinkShuffleByPk =
+                config.getConfiguration().get(ExecutionConfigOptions.TABLE_EXEC_SINK_KEYED_SHUFFLE);
+        boolean sinkKeyBy = false;
+        switch (sinkShuffleByPk) {
+            case NONE:
+                break;
+            case AUTO:
+                sinkKeyBy = inputInsertOnly && sinkParallelism != inputParallelism;
+                break;
+            case FORCE:
+                // single parallelism has no problem
+                sinkKeyBy = sinkParallelism != 1 || inputParallelism != 1;
```

Comment:
@lincoln-lil Could you explain `inputParallelism != 1`? I know it is a corner case but isn't the FORCE mode weaker than AUTO for insert-only streams? AUTO allows e.g. sinkParallelism=4 and inputParallelism=1.

Reply:
@twalthr if 'AUTO allows e.g. sinkParallelism=4 and inputParallelism=1' than FORCE also allows it. FORCE mode isn't weaker than AUTO.
This condition replace the previous one: '!(sinkParallelism == 1 && inputParallelism == 1)' via Jingsong's advice.
It seems the previous version still easier to understand..

Reply:
I was torn about this condition because I originally thought we shouldn't specialize the single parallelism case. But considering that there are a lot of single parallelism jobs in stream computing, it is worth optimizing to remove keyBy for single parallelism.

Reply:
Thanks for the explanation. I somehow misread `||` vs `&&`. Optimizing for the single parallelism case makes sense to me even in FORCE mode.